iceoryx_utils/platform/mac/source/semaphore.cpp

```diff
@@ -0,0 +1,253 @@
+// Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iceoryx_utils/platform/semaphore.hpp"
+#include "iceoryx_utils/platform/time.hpp"
+
+#include <chrono>
+#include <cstdarg>
+#include <cstdlib>
+#include <iostream>
+#include <thread>
+
+static void deleteValue(std::atomic<int>*& value) noexcept
+{
+    if (value != nullptr)
+    {
+        delete value;
+        value = nullptr;
+    }
+}
+
+iox_sem_t::iox_sem_t() noexcept
+    : m_value{new std::atomic<int>}
+{
+}
+
+iox_sem_t::iox_sem_t(iox_sem_t&& rhs) noexcept
+{
+    *this = std::move(rhs);
+}
+
+iox_sem_t::~iox_sem_t()
+{
+    deleteValue(m_value);
+}
+
+iox_sem_t& iox_sem_t::operator=(iox_sem_t&& rhs) noexcept
+{
+    if (this != &rhs)
+    {
+        deleteValue(m_value);
+        m_handle = rhs.m_handle;
+        m_hasPosixHandle = rhs.m_hasPosixHandle;
+        m_value = rhs.m_value;
+        rhs.m_value = nullptr;
+    }
+    return *this;
+}
+
+int iox_sem_getvalue(iox_sem_t* sem, int* sval)
+{
+    *sval = sem->m_value->load();
+    return 0;
+}
+
+int iox_sem_post(iox_sem_t* sem)
+{
+    int retVal{0};
+    if (sem->m_hasPosixHandle)
+    {
+        retVal = sem_post(sem->m_handle.posix);
+        if (retVal == 0)
+        {
+            (*sem->m_value)++;
+        }
+    }
+    else
+    {
+        // dispatch semaphore always succeed
+        dispatch_semaphore_signal(sem->m_handle.dispatch);
+        (*sem->m_value)++;
+    }
+    return retVal;
+}
+
+int iox_sem_wait(iox_sem_t* sem)
+{
+    int retVal{0};
+    if (sem->m_hasPosixHandle)
+    {
+        retVal = sem_wait(sem->m_handle.posix);
+    }
+    else
+    {
+        // dispatch semaphore always succeed
+        dispatch_semaphore_wait(sem->m_handle.dispatch, DISPATCH_TIME_FOREVER);
+    }
+
+    sem->m_value->fetch_sub((retVal == 0) ? 1 : 0);
```

Comment:
how about putting this in a function like this?
```
void foo(atomic<??>& value, int retval)
{
    if(retval == 0)
    {
        value.fetch_sub(1, std::memory_order_relaxed);
    }
}
```
you should also probably put it into an anonymous namespace

Reply:
Resolved by using value--